Bob's Shake

teiknsmith opened the pull request:
There are still two main things to do, but I would like second eyes on this before doing the final testing, and fixing the locks log will need to happen after the merge, so this is ready enough for a review.

teiknsmith committed:
Create entry point for Bob's Shake

teiknsmith committed:
Add bobs_shake.py

SeaSaltTree committed:
Started trying to mmake event class for locking and unlocking

Started making a function to read in the .csv file and make events out of each row in the file

SeaSaltTree committed:
Worked on grading stats worker class

Finished up work_events constructor and started assignments

SeaSaltTree committed:
fixed style

seyfdall committed:
Imported get_lock_log_path and added some tests

SeaSaltTree committed:
Added to debugging functionality

Made a debugging function to write out to a file and left some comments on stuff that is broken

SeaSaltTree committed:
Tried debugging WorkEvent

When initializing work event there seems to be errors with assigning each attribute

SeaSaltTree committed:
Debugged WorkEvent

Attributes from each row for each event are not being entered properly

SeaSaltTree committed:
Debugged WorkEvent

Attributes from rows aren't being assigned correctly to their variables

teiknsmith committed:
Finish reading native stats in Bob's Shake

teiknsmith committed:
Move classes to top level of module

teiknsmith committed:
Prepare for analyzing events

teiknsmith committed:
Add reading stats from helpqueue

teiknsmith committed:
Create TA Data and Storage

teiknsmith committed:
Remove Display Name from TA model
It really isn't needed - the name from the queue will be a
good enough display name if we really need one

teiknsmith committed:
Add Name Validator
Not fully debugged yet
Matching TA names from the queue vs from zygrader native
will allow for better analysis later

teiknsmith committed:
Simplify Name Validation
Also, some of the simple FIXMEs

teiknsmith committed:
Add queue events when assigning to TAs

teiknsmith committed:
Show events grouped by TA (for debugging)

teiknsmith committed:
Add step to analyze events

teiknsmith committed:
Implement deduplication of nested events (pair programming)

teiknsmith committed:
Rename entry point for Bob's Shake
Because why not. Shake is a verb too.

teiknsmith committed:
Add uniq identifier to WorkEvents for what item is involved
For labs, the student and lab name
For emails, the student name
For helps, a special id (as they are generated in pairs)

teiknsmith committed:
Begin analyzing event streams

teiknsmith committed:
Continue analysis

teiknsmith committed:
Cleanup old group analysis method, prepare for new

teiknsmith committed:
Add prompt to queue file entry

teiknsmith committed:
Update global TODO

teiknsmith committed:
Finish analysis of event streams
Largely by reducing the requirements, but the extra items
were determined to be unnecessary

teiknsmith committed:
Add outputting the stats

teiknsmith committed:
Remove debugging code

teiknsmith committed:
Fix import error

teiknsmith committed:
Cleanup more debug/errata

teiknsmith committed:
Manually parse durations from queue data
In checking when errors occured, it turns out this is more robust

teiknsmith committed:
Present errors generated by Queue Data
Because this data comes from an outside source, it is more likely
 to cause errors. Hopefully this doesn't actually happen, I've
 tested with a lot of queue data, but in case it does

teiknsmith committed:
Catch all errors created from reading Queue Data

teiknsmith committed:
Cleanups and preparation for review

natecraddock commented:
Could you merge/rebase master?

natecraddock commented on zygrader/bobs_shake.py:
```diff
@@ -0,0 +1,394 @@
+#TODO: Check this against the old Bob's Second
+#TODO: run my 'history fixer' on the lock log
```
Space after `#`

natecraddock commented on zygrader/bobs_shake.py:
```diff
@@ -0,0 +1,394 @@
+
+        uniq_item = event_type + student_name + assignment_name
+
+        # FIXME: ta_netid is a lab name if student has a comman in their name
+        #  (this is somewhat fixed in data/lock.py by using csv to write data,
+        #   but old data might still have commas)
+        #  one possible fix (and what I'm thinking I'll do) is to modify the
+        #   the lock log files outside of zygrader,
+        #   just with some separate script, to keep this area clean
+
```
The comment `# ` spacing here...

natecraddock reviewed (commented):
Perhaps explain in the source code what "Shake" means

At a high level this code looks really good! I'll take a closer look tomorrow

teiknsmith committed:
Cleanup block comments

teiknsmith committed:
Replace was_canceled function calls with canceled member access

teiknsmith committed:
Make Bob's Shake More Friendly as a Module
Add docstrings for the module and `shake` function
Use _leading_underscore for classes and functions that don't need export

teiknsmith commented:
> Could you merge/rebase master?

yeah, I probably should have done that. it has been done now

> Perhaps explain in the source code what "Shake" means

added some changes to help with that too

natecraddock commented on zygrader/bobs_shake.py:
```diff
@@ -0,0 +1,422 @@
+    worker = _StatsWorker()
+    Step = namedtuple('Step', ['interactive', 'msg', 'func'])
+    steps = [
+        Step(True, "start time", lambda: worker.select_start_time()),
+        Step(True, "end time", lambda: worker.select_end_time()),
+        Step(True, "select helpqueue csv",
```
Probably use title case or at least capitalize the messages here

teiknsmith replied:
for the interactive steps, the msg is never presented, so I was purposefully leaving these lower case. they only really serve to help in reading the code, and I'm realizing the func name does this well enough, so I'll actually just make them empty I think

natecraddock commented on zygrader/bobs_shake.py:
```diff
@@ -0,0 +1,422 @@
+
+        uniq_item = event_type + student_name + assignment_name
+
+        # FIXME: ta_netid is a lab name if student has a comman in their name
```
comman -> comma

teiknsmith replied:
this comment will be deleted in the final steps of finishing this when I csv-ify the lock log file, but fair catch

natecraddock commented on zygrader/bobs_shake.py:
```diff
@@ -0,0 +1,422 @@
+        time_stamp = datetime.datetime.strptime(time_stamp_str,
+                                                "%Y-%m-%dT%H:%M:%S.%f")
+
+        # The old lock format did not have an event_type field
```
```suggestion
        # The old lock format did not have an event_type field
        # (all locks were for labs)
        is_old_format = len(row) == 5
        if is_old_format:
            row.insert(1, "LAB")

        event_type = row[1]
        student_name = row[2]
        assignment_name = row[3]
        ta_netid = row[4]
        
        ...
        
        lock_type = row[5]
```

I am not sure if keeping `row` intact matters, but this is more readable to me

teiknsmith replied:
it doesn't matter if row stays intact, and I agree. I'll change this

natecraddock commented on zygrader/bobs_shake.py:
```diff
@@ -0,0 +1,422 @@
+            if event.is_begin:
+                unmatched_depth += 1
+                new_list.append(event)
+            else:
+                pass
```
Remove

teiknsmith replied:
I had that to make it clear that nothing needs to happen there if the event is not begin. but a comment would probably be even better

natecraddock commented on zygrader/data/__init__.py:
```diff
@@ -122,3 +122,38 @@ def write_class_sections(class_sections):
+
+    path = SharedData.get_ta_data()
+    with open(path, "w") as _file:
+        json.dump(tas_json, _file, indent=2)
```
End files with trailing newline

natecraddock reviewed (commented):
Looks good to me. I'd like to take one last look before you commit (since you implied there are a few more things to finish)

natecraddock commented:
I should mention, there were a few files that were just whitespace cleanups - that can go in now (no need to keep in this branch and PR)

teiknsmith committed:
Cleanup Steps
Don't include a message if there is no need for it
Remove lambda wrapping around functions

teiknsmith committed:
Fix typo 'comman'

teiknsmith committed:
Cleanup handling of old lock log format
Modify the row to make it consistent with the new format,
rather than using less readable logic to work around it

teiknsmith committed:
Replace empty else with explanatory comment

teiknsmith committed:
Add trailing newline to end of file

teiknsmith committed:
Refactor Deduplication
Only "remove" events if they are completely sandwiched by another pair

teiknsmith committed:
Add Default Times for Time Selecting

teiknsmith committed:
Remove TODO of Checking Against Bob's Second
Any changes between the results of Bob's Second and Bob's Shake
are negligible or improvements

teiknsmith commented:
this is now in its (hopefully) final form, with the exception of one TODO which is just a reminder for after this gets added to an update

> I'd like to take one last look before you commit (since you implied there are a few more things to finish)

so this can be done now

natecraddock reviewed (approved):
Perhaps do a squash commit so the log is a bit more clear and not full of various back-and-forth changes

Looks good!
